rustls/src/msgs/deframer.rs

```diff
@@ -87,9 +91,22 @@ impl MessageDeframer {
                 }
             };
 
-            // Return CCS messages immediately without decrypting.
+            // Return CCS messages and early plaintext alerts immediately without decrypting.
             let end = start + rd.used();
-            if m.typ == ContentType::ChangeCipherSpec && self.joining_hs.is_none() {
+            let type_is_ccs = m.typ == ContentType::ChangeCipherSpec;
+            let has_decrypted = record_layer.has_decrypted();
+            let type_is_alert = m.typ == ContentType::Alert;
+            let payload_is_alert = m.payload.0.len() <= 2;
+            let version_is_tls13 = matches!(negotiated_version, Some(ProtocolVersion::TLSv1_3));
```

Comment:
From first principles, why are we limiting this to TLS 1.3 again?

Reply:
TLS1.2 has ChangeCipherSpec messages which make it unambiguous when the keying changes (including when transitioning from plaintext to encrypted). TLS1.3 did away with those[^1] which means the ambiguity of whether an alert will be encrypted is specific to TLS1.3.

[^1]: they can still be sent for "middlebox compatibility mode", but they have no meaning and cannot be relied on

Reply:
I will try to capture more of this in the PR description and comments.